File: src/main/java/org/tindalos/principle/infrastructure/service/jdepend/PackageListFactory.java
package org.tindalos.principle.infrastructure.service.jdepend;

import java.util.Collection;
import java.util.List;

import jdepend.framework.JavaPackage;

import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.detector.core.PackageSorter;

import com.google.common.collect.Lists;

public class PackageListFactory {

	private final PackageFactory packageFactory;
	private final PackageSorter packageSorter;

	public PackageListFactory(PackageFactory packageFactory, PackageSorter packageSorter) {
		this.packageFactory = packageFactory;
		this.packageSorter = packageSorter;
	}

	public List<Package> build(Collection<JavaPackage> analyzedPackages) {

		List<Package> result = Lists.newArrayList();
		for (JavaPackage javaPackage : analyzedPackages) {
			if (packageFactory.isRelevant(javaPackage)) {
				Package aPackage = packageFactory.transform(javaPackage);
				result.add(aPackage);
			}
		}
		return packageSorter.sortByName(result);
	}

}

File: src/main/java/org/tindalos/principle/infrastructure/service/jdepend/PackageListFactory.scala
package org.tindalos.principle.infrastructure.service.jdepend

import org.tindalos.principle.domain.core.PackageSorter
import jdepend.framework.JavaPackage
import scala.collection.immutable.List
import org.tindalos.principle.domain.core.Package

object PackageListFactory {

  def buildPackageListFactory(packageFactory:PackageFactory, packageSorter:PackageSorter) =

    (analyzedPackages: List[JavaPackage]) => {

      val unsortedList = analyzedPackages
        .filter(packageFactory.isRelevant(_))
        .map(packageFactory.transform(_))

      packageSorter.sortByName(unsortedList)
    }
}